Add retries to advanced audit e2e tests.

loburm opened the pull request:
Fixes #56683

```release-note
NONE
```

loburm committed:
Add retries to advanced audit e2e tests.

loburm commented:
/assign @crassirostris @sttts

ericchiang commented on test/e2e/auth/audit.go:
```diff
@@ -653,24 +656,33 @@ func expectAuditLines(f *framework.Framework, expected []auditEvent) {
+	for i := 0; i < retries; i++ {
+		stream, err := f.ClientSet.CoreV1().RESTClient().Get().AbsPath("/logs/kube-apiserver-audit.log").Stream()
+		defer stream.Close()
+		framework.ExpectNoError(err, "could not read audit log")
```
Don't you want

```
if err != nil {
    continue
}
```

loburm replied:
Done.

ericchiang commented on test/e2e/auth/audit.go:
```diff
@@ -653,24 +656,33 @@ func expectAuditLines(f *framework.Framework, expected []auditEvent) {
-		if _, found := expectations[event]; found {
-			expectations[event] = true
+	for i := 0; i < retries; i++ {
+		stream, err := f.ClientSet.CoreV1().RESTClient().Get().AbsPath("/logs/kube-apiserver-audit.log").Stream()
```
This shadows the `err` variable you've defined outside the loop.

ericchiang replied:
Maybe put this in a closure? You'd avoid shadowing and would execute the defers immediately. 

```
err = func() error {
    // execute logic
}()
if err == nil {
    break
}
```

loburm replied:
Thank you for suggestion. Done.

ericchiang commented on test/e2e/auth/audit.go:
```diff
@@ -653,24 +656,33 @@ func expectAuditLines(f *framework.Framework, expected []auditEvent) {
+			event, err := parseAuditLine(line)
+
+			if err != nil {
+				continue
```
log this error?

loburm replied:
I'm afraid that if something is really broken in audit log then it will just spam us with error messages (in 5k node clusters there are hundreds of thousand log lines, we don't want to see error message for each one).

porridge commented:
I'm not sure this change is sufficient. Note that the audit log gets rotated and compressed (probably due to its size) on the 5k node cluster test: http://gcsweb.k8s.io/gcs/kubernetes-jenkins/logs/ci-kubernetes-e2e-gce-scale-correctness/40/artifacts/gce-scale-cluster-master/

loburm commented:
@ericchiang I have addressed your comments. PTAL.
@porridge In scalability test rotation happens when log achieves 5GB (https://github.com/kubernetes/test-infra/blob/master/jobs/env/ci-kubernetes-e2e-gce-large-correctness.env#L10).
 I think that chances that we will hit rotation two times in row is miserable.

porridge commented:
@loburm I think it's enough to hit one rotation for the test to fail

crassirostris commented:
@porridge Do we have some estimates for the probability?

BTW, that's something that [Tim's FR](https://github.com/kubernetes/kubernetes/issues/53455) addresses

loburm commented:
@crassirostris hope my math is correct:
Average audit log line is near 1300bytes long, which means that we are going to get near 4M of logs written. Unfortunately I don't remember how many log lines have been generated during this exact test, but it was less than 10k. Even if we assume that's a true, then chances that log is going to be rotated exactly during this test run is 1/400. Chances that it happens two times in row: 1/160000. I think that should be acceptable, considering that we run this test few times per week.

I have bigger concerns that we are going to have network problems when such huge files are transferred.

crassirostris commented:
I think the main point is that the FR I mentioned has actually a higher priority than anticipated. I'll update it with a use case

porridge commented:
Regarding the probability calculations: I don't think we should be
designing tests assuming they will be retried. 1/400 sounds like an
unacceptably pretty high probability for two reasons:
1. these test jobs are extremely costly, in multiple dimensions (monetary,
time, frequency), we need to make sure we avoid flakes
2. there are a few hundred tests in this suite, if every test author
assumed 1/400 failure rate is fine, then we'd never have a single
successful run. Please try to be a responsible citizen.

porridge commented:
As an aside, I'm not sure I understand why this problematic test was
introduced into this suite that late in the release cycle? According to
https://k8s-testgrid.appspot.com/sig-scalability-gce#gce-scale-correctness
it was not run on Nov 25th, and the freeze date was Nov 22nd. Was this a
result of some late exception?

crassirostris commented:
@porridge Wait, I've just realized that we shouldn't even have audit logging at scale right now. There's an ongoing problem that will not be fixed until 1.10, that audit logging in GCE environment is decreasing the performance because of synchronous disk writes (more context in https://github.com/kubernetes/kubernetes/issues/53006)

Is there a way to run this test in the default suite, but not in the 5k nodes test?

porridge commented:
Sure, add a setting similar to
https://github.com/kubernetes/test-infra/blob/master/jobs/env/ci-kubernetes-e2e-gce-scale-performance.env#L28-L29
to the -correctness.env file

crassirostris commented:
However, will it disable the test? Just disabling audit logging with audit logging e2e in the suite will cause it to fail with probability 1

porridge commented:
Ouch, right.
To disable the test, add [DisabledForLargeClusters] to test title. Please
add a reference to tracking bug for un-disabling in a comment nearby in
that case.

crassirostris commented:
Great, thanks!

@loburm, could you please do this?

loburm commented:
Marcin thanks for the pointer. I have created a new PR for this: #56791

porridge commented:
We should probably still disable the feature altogether in the env file, no?

crassirostris commented:
Yes, we should. @loburm, can I ask you to make another PR?

thockin commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/56705)
<!-- Reviewable:end -->
